### ISSUE TITLE
[Bug]: Deadlock in LocalCPUBackend.allocate() when use_hot=False and staging buffer is full

### ISSUE THREAD

**Issue by ultranationalism** [LMCache/LMCache#2942]
**Label**
bug

**Describe the bug**

When `local_cpu=False` (i.e., `use_hot=False`), `LocalCPUBackend.allocate(busy_loop=True)` enters an infinite busy-wait loop if the CPU staging buffer is full. The eviction path is entirely gated by `if self.use_hot:`, so when `use_hot=False`, no eviction is ever attempted — the loop just calls `time.sleep(0.1)` and retries `memory_allocator.allocate()` forever.

This is triggered in production by `LocalDiskBackend.load_bytes_from_disk()` (`local_disk_backend.py:579`), which calls `allocate()` with the default `busy_loop=True`. The symptom is the engine becoming completely unresponsive with this warning flooding the logs:

```
LMCache WARNING: No eviction candidates found in local cpu backend.
Local cpu memory is under pressure. Waiting for 0.1 seconds before retrying.
```

**Configuration that triggers this:**
```yaml
chunk_size: 256
local_cpu: false
max_local_cpu_size: 5        # GB — staging buffer for disk offload
local_disk: "file:///data/lmcache/"
max_local_disk_size: 200
```

Per the documentation, `max_local_cpu_size` should be set even when `local_cpu=false`, as the CPU memory pool serves as a staging buffer for disk I/O. This is a valid and intended configuration.

**Root Cause Analysis**

The core issue is in `LocalCPUBackend.allocate()` (`local_cpu_backend.py:533-580`):

```python
while True:
    wait_other_requests = True
    if self.use_hot:              # ← skipped when use_hot=False
        evict_keys = self.cache_policy.get_evict_candidates(...)
        if evict_keys:
            wait_other_requests = False
            self.batched_remove(evict_keys, force=False)

    if wait_other_requests:       # ← always True when use_hot=False
        if not busy_loop:
            break                 # busy_loop=False: safe, returns None
        time.sleep(0.1)           # busy_loop=True: spins forever!

    memory_obj = self.memory_allocator.allocate(...)
    if memory_obj is not None:
        break
```

The same pattern exists in `batched_allocate()` (`local_cpu_backend.py:643-706`).

**Contributing factors that form a deadlock chain:**

1. **No active eviction path** — When `use_hot=False`, `hot_cache` is empty (because `batched_submit_put_task` returns early at line 193-194). The eviction logic only looks at `hot_cache`, so there is nothing to evict even though the staging buffer is full of in-flight objects.

2. **Staging buffer occupied by in-flight put tasks** — `submit_put_task()` calls `ref_count_up()` (line 344) before submitting an async disk write. The buffer is held until `async_save_bytes_to_disk()` completes and calls `ref_count_down()` (line 526). Slow disk I/O prolongs buffer occupation.

3. **Put tasks starved by priority scheduling** — `async_save_bytes_to_disk` is submitted with priority 2 (lowest), while prefetch tasks have priority 0 (highest). Under load, put completions (and their `ref_count_down()` calls) are delayed.

4. **New stores fail silently** — `allocate_and_copy_objects()` (storage_manager.py:89-94) uses `busy_loop=False`. When the buffer is full, store allocation returns `None` and breaks — no new data is written, so no new `ref_count_down()` calls will free buffer space.

5. **Cascading failure:**
```
Executor busy (prefetch priority > put)
  → put tasks delayed → ref_count_down() delayed
  → staging buffer stays full
  → new put allocations fail (busy_loop=False, break)
  → no new writes → no new buffer releases
  → get_blocking's busy_loop=True spins forever
  → engine thread blocked → all requests stall
```

**Why `use_hot=True` is not affected:**

With `use_hot=True`, `submit_put_task()` inserts `memory_obj` into `hot_cache` (line 159). When `allocate()` needs space, it synchronously evicts LRU entries from `hot_cache` via `batched_remove()` — this is a local, lock-protected operation that does not depend on the async executor, disk I/O completion, or any external scheduling. It breaks the entire deadlock chain.

**To Reproduce**

```python
import threading
import torch

from lmcache.utils import CacheEngineKey
from lmcache.v1.config import LMCacheEngineConfig
from lmcache.v1.memory_management import MemoryFormat, MixedMemoryAllocator
from lmcache.v1.metadata import LMCacheMetadata
from lmcache.v1.pin_monitor import PinMonitor
from lmcache.v1.storage_backend.local_cpu_backend import LocalCPUBackend

CHUNK_SHAPE = torch.Size([2, 256, 8, 128])  # ~1 MB at bf16
CHUNK_DTYPE = torch.bfloat16
CHUNK_FMT = MemoryFormat.KV_2LTD
POOL_BYTES = 3 * 1024 * 1024  # 3 MB, fits ~3 chunks


def make_backend(use_hot: bool):
    config = LMCacheEngineConfig.from_defaults(
        chunk_size=256, local_cpu=use_hot, lmcache_instance_id="deadlock_test",
    )
    metadata = LMCacheMetadata(
        model_name="test", world_size=1, local_world_size=1,
        worker_id=0, local_worker_id=0,
        kv_dtype=CHUNK_DTYPE, kv_shape=(4, 2, 256, 8, 128),
    )
    PinMonitor.GetOrCreate(config)
    allocator = MixedMemoryAllocator(POOL_BYTES)
    backend = LocalCPUBackend(
        config=config, metadata=metadata,
        memory_allocator=allocator, dst_device="cpu",
    )
    return backend, allocator


def test_scenario(use_hot: bool):
    backend, allocator = make_backend(use_hot)
    try:
        # Fill pool
        held = []
        for i in range(100):
            obj = backend.allocate(CHUNK_SHAPE, CHUNK_DTYPE, CHUNK_FMT,
                                   eviction=False, busy_loop=False)
            if obj is None:
                break
            if use_hot:
                key = CacheEngineKey(model_name="test", world_size=1,
                                     worker_id=0, chunk_hash=i, dtype=CHUNK_DTYPE)
                backend.submit_put_task(key, obj)
                obj.ref_count_down()
            held.append(obj)

        # Try allocate like load_bytes_from_disk does
        result_box = [None]
        def try_alloc():
            result_box[0] = backend.allocate(CHUNK_SHAPE, CHUNK_DTYPE, CHUNK_FMT,
                                              eviction=True, busy_loop=True)
        t = threading.Thread(target=try_alloc, daemon=True)
        t.start()
        t.join(timeout=3.0)

        label = f"use_hot={use_hot}"
        if t.is_alive():
            print(f"  {label}: DEADLOCK (thread stuck after 3s)")
        else:
            print(f"  {label}: OK (eviction worked)")
            if result_box[0] is not None:
                result_box[0].ref_count_down()

        if not use_hot:
            for obj in held:
                obj.ref_count_down()
    finally:
        allocator.close()
        PinMonitor.DestroyInstance()


if __name__ == "__main__":
    test_scenario(use_hot=False)  # DEADLOCK
    test_scenario(use_hot=True)   # OK
```

**Expected output:**
```
  use_hot=False: OK
  use_hot=True: OK
```

**Actual output:**
```
  use_hot=False: DEADLOCK (thread stuck after 3s)
  use_hot=True: OK (eviction worked)
```

**Expected behavior**

When `use_hot=False` and the staging buffer is full, `allocate()` should not spin indefinitely. It should either:
- Return `None` after a bounded number of retries (and callers like `load_bytes_from_disk` should handle this gracefully instead of asserting), or
- Track staging buffer objects in an evictable structure so the eviction path works regardless of `use_hot`, or
- Force `busy_loop=False` when `use_hot=False`, since there is no active reclamation mechanism to make busy-looping productive.

**Affected code:**
- `lmcache/v1/storage_backend/local_cpu_backend.py`: `allocate()` L533-580, `batched_allocate()` L643-706
- `lmcache/v1/storage_backend/local_disk_backend.py`: `load_bytes_from_disk()` L579 (calls allocate with default `busy_loop=True`), `get_blocking()` L404

**Environment:**
- LMCache version: v0.4.2 (commit 996da472)
- OS: Linux
- Context length: 80K tokens
- Model: serving with vLLM

**Additional context**

The design comment in `allocate()` (lines 503-513) states:
> *busy_loop should only be used for retrieve ... one retrieve at a time (okay to busy loop because stores will clear)*

This assumption holds when `use_hot=True` (eviction can synchronously free space), but breaks completely when `use_hot=False` — stores may not clear because they also fail to allocate, and there is no eviction mechanism to break the cycle.

**Comment by ianliuy** [comment]
Thanks for the detailed report — the root cause analysis is spot-on.

I've traced through the code and confirmed the deadlock chain:

1. `batched_submit_put_task()` returns early at L193-194 when `use_hot=False`, so `hot_cache` is always empty
2. `allocate()` L536 gates the entire eviction block on `if self.use_hot:` — when `False`, `wait_other_requests` is never cleared
3. With `busy_loop=True` (the default), the loop spins on `time.sleep(0.1)` forever
4. The same structural bug exists in `batched_allocate()` L643-706

**Fix I'm working on (Option 2 from the issue):**

Override `busy_loop` to `False` internally when `use_hot=False`, since the docstring's own precondition ("stores will clear") is structurally impossible in this configuration:

```python
# allocate() and batched_allocate() — before the while True loop
if not self.use_hot:
    busy_loop = False  # no eviction mechanism exists; busy-looping is guaranteed futile
```

Also fix `load_bytes_from_disk()` to replace the `assert memory_obj is not None` with graceful `None` handling and a warning log (matching the pattern already used in `batched_get_non_blocking()`), so the failure degrades to a cache miss rather than an assert crash.

This is a minimal fix: ~8 lines changed, zero behavioral change when `use_hot=True`.

Working on a PR now.

**Comment by ianliuy** [comment]
Hi @ultranationalism, thanks for the detailed root cause analysis  it made the fix straightforward to implement.

## Fix Summary

PR: https://github.com/LMCache/LMCache/pull/new/fix/issue-2942 (opening shortly targeting `dev`)

### Root cause confirmed
`LocalCPUBackend.allocate()` and `batched_allocate()` busy-loop forever when `use_hot=False` because the only code path that clears `wait_other_requests` is gated by `if self.use_hot:`. Since `batched_submit_put_task()` returns early without inserting into `hot_cache` when `use_hot=False`, eviction is structurally impossible regardless of how long the loop spins.

### Changes (2 files, +23 lines, -1 line)

**`local_cpu_backend.py`**  core fix in both `allocate()` and `batched_allocate()`:
```python
# When use_hot=False, hot_cache is always empty so eviction is
# structurally impossible. Override to fail-fast instead of spinning.
if not self.use_hot:
    busy_loop = False
```
This is a complete no-op when `use_hot=True`. Aligns with the existing docstring: *"busy_loop works because stores will clear"*  they provably cannot when `use_hot=False`.

**`local_disk_backend.py`**  two fixes:
1. `load_bytes_from_disk()`: replaced `assert memory_obj is not None` with a warning log + `return None`, so the caller gets a cache miss instead of an assertion crash or infinite hang.
2. `batched_get_non_blocking()`: fixed a pre-existing `disk_lock` leak on the allocation-failure path (early return at L475 was missing `self.disk_lock.release()`). This lock leak would have permanently deadlocked all subsequent disk operations  it was previously unreachable but becomes reachable under memory pressure after the above fixes.

### Follow-up needed (separate issue)
`get_blocking()` calls `cache_policy.update_on_hit()` *before* confirming the data is loaded  a phantom hit is recorded when `load_bytes_from_disk()` returns `None`. This is pre-existing but now reachable. Would be good to track in a separate issue.

**Comment by amirhnajafiz** [comment]
@ianliuy @ultranationalism Hi folks, I wanted to share an update. I was able to reproduce the issue even when the hot cache is enabled (`local_cpu=True`). I’m using vLLM (v0.15.0) with LMCache (0.4.1).

My LMCache configuration:

```
lmcache.yaml: |
    internal_api_server_enabled: True
    internal_api_server_port_start: 6999

    chunk_size: 16
    save_unfull_chunk: false
    save_decode_cache: false
    min_retrieve_tokens: 0

    local_cpu: True
    max_local_cpu_size: 0.1

    extra_config: {'use_odirect': True}
    local_disk: "file:///mnt/localstorage/cache"
    max_local_disk_size: 20.0
```

I limited my GPU tier to 1.2 GB of VRAM for KV Cache to ensure offloading occurs. I also minimized the local cpu size. Then I used the `meta-llama/Llama-3.2-1B-Instruct` model with the ShareGPT benchmark (200 requests, 5 concurrent requests).

On the first run, all KV data goes through the store operation, so nothing unusual happens. I only observe about 7.5 GB used on local disk. However, when I rerun the benchmark with the same parameters, I eventually hit a deadlock.

```
(EngineCore_DP0 pid=142) [2026-04-23 17:39:35,970] LMCache WARNING: No eviction candidates found in local CPU backend. Local CPU memory is under pressure. Waiting for 0.1 seconds before retrying. (local_cpu_backend.py:569:lmcache.v1.storage_backend.local_cpu_backend)
(EngineCore_DP0 pid=142) [2026-04-23 17:39:36,070] LMCache WARNING: No eviction candidates found in local CPU backend. Local CPU memory is under pressure. Waiting for 0.1 seconds before retrying. (local_cpu_backend.py:569:lmcache.v1.storage_backend.local_cpu_backend)
```

This issue occurs specifically when attempting to load a large number of tokens at once:

```
2026-04-23 13:15:57.651	
(EngineCore_DP0 pid=142) [2026-04-23 17:15:57,651] LMCache INFO: Reqid: cmpl-8e5d77b772713c10-0-917ec873, Total tokens: 5436, Inference Engine computed tokens: 0, LMCache hit tokens: 5424, need to load: 5424 (vllm_v1_adapter.py:1316:lmcache.integration.vllm.vllm_v1_adapter)
2026-04-23 13:15:57.768	
(EngineCore_DP0 pid=142) [2026-04-23 17:15:57,767] LMCache WARNING: No eviction candidates found in local CPU backend. Local CPU memory is under pressure. Waiting for 0.1 seconds before retrying. (local_cpu_backend.py:569:lmcache.v1.storage_backend.local_cpu_backend)
2026-04-23 13:16:01.673	
(APIServer pid=12) INFO:     192.168.126.223:39710 - "GET /metrics HTTP/1.1" 200 OK
```

**Comment by ianliuy** [comment]
Fixed the CI failure in PR #3006.

**Root cause**: The test `test_load_bytes_from_disk_returns_none_on_pool_exhaustion` used `capsys` to capture stderr, but `logging.StreamHandler()` grabs `sys.stderr` at handler-creation time. When `capsys` later replaces the Python `sys.stderr` object, the handler still writes to the original file descriptor — so `capsys` never sees the output.

**Fix**: Switched from `capsys` to `capfd` (fd-level capture via `dup2`), which captures all writes to fd 2 regardless of Python object references. 3-line change, zero logic impact.

Pushed to PR #3006: https://github.com/LMCache/LMCache/pull/3006
